### DISCUSSION TITLE
[Help Needed] Using the YARP as HTTP forwarder for https url requests

### DISCUSSION THREAD

**Discussion by ghost** [dotnet/yarp#1637] [Q&A] [answered]
Team, 
We are using the YARP HTTP Frowarder abilities as a forward proxy server to one of our internal service.  Our proxy service is listening on http (e.g http://localhost:3030). All the http url requests via proxy work all good however, when I try to use the "https" requests then the calls fails and when we debug further inside the YARP endpoint mapping all the time httpContext.Request.IsHttps comes as false. Any inputs to further troubleshoot this issue would of great help. Thanks.

```c#
app.UseEndpoints(endpoints =>
{
    endpoints.Map("/{**catch-all}", async httpContext =>
    {
        try
        {
            var prefix = httpContext.Request.IsHttps ? "https://" : "http://";
            var destinationAddress = prefix + httpContext.Request.Host;
            var error = await forwarder.SendAsync(httpContext, destinationAddress,
                httpClient, requestConfig, transformer);
            // Check if the operation was successful
            if (error != ForwarderError.None)
            {
                var errorFeature = httpContext.GetForwarderErrorFeature();
                var exception = errorFeature.Exception;
            }
        }
        catch (Exception e)
        {
            Console.WriteLine(e);
        }
    });
});
```

**Comment by greenEkatherine** [comment, marked as answer]
Did you configure an endpoint for https like here https://docs.microsoft.com/en-us/aspnet/core/fundamentals/servers/kestrel/endpoints?view=aspnetcore-6.0#configureiconfiguration ?

**Reply by suntsu42** [reply]
I do have the very same problem.

When i use http://localhost:666/test/ the redirect works as expected. But when i enter http://localhost:666/test (Without trailing slash), the browser is redirected to http://localhost:777/test/ which is the url under which the content is actually hosted (via nginx). 

Is there any setting in yarp in order to avoid this problem?



code: 
```csharp
using Microsoft.AspNetCore.Authentication.OpenIdConnect;
using Microsoft.Identity.Web;
using NLog;
using System.Net;

const string reverseProxy = "ReverseProxy";
const string urlConfig = "Url";

//Create the logger instance based on NLog.config
Logger logger = LogManager.GetCurrentClassLogger();
var builder = WebApplication.CreateBuilder(args);

builder.Services.AddReverseProxy().LoadFromConfig(builder.Configuration.GetSection(reverseProxy));
var app = builder.Build();
app.MapReverseProxy();
app.Run(builder.Configuration[urlConfig]);
```

config(appsettings.json):
```json
{
  "Logging": {
    "LogLevel": {
      "Default": "Information",
      "Microsoft.AspNetCore": "Warning"
    }
  },
  "AllowedHosts": "*",
  "ReverseProxy": {
    "Routes": {
      "route1": {
        "ClusterId": "cluster1",
        "Match": {
          "Path": "{**catch-all}"
        }
      }
    },
    "Clusters": {
      "cluster1": {
        "Destinations": {
          "destination1": {
            "Address": "http://localhost:777"
          }
        }
      }
    }
  },
  "Url": "http://localhost:666"

}
```


Edit: What actually happens is that the request for http://localhost:666/test should be redirected to http://localhost:666/test/index.html but isn't. 
Nginx returns a 301 in this case. 
Can this be fixed in yarp?
![image](https://user-images.githubusercontent.com/1768336/162926608-746a997d-c4e8-45b9-b993-65ff29f0a060.png)

**Reply by MihaZupan** [reply]
@suntsu42 The behavior with/without a trailing slash is a different issue. Please see https://github.com/microsoft/reverse-proxy/discussions/1633#discussioncomment-2501763

**Reply by suntsu42** [reply]
> @suntsu42 The behavior with/without a trailing slash is a different issue. Please see [#1633 (reply in thread)](https://github.com/microsoft/reverse-proxy/discussions/1633#discussioncomment-2501763)

This explains and solves my issue. Thanks

**Reply by sqlnew** [reply]
@rameshnlg 
Has this problem finally been resolved?

I use YARP HTTP Forwarder as a forward proxy, which means specifying it as a proxy server in the browser, such as setting the proxy server IP port to 127.0.0.1:5000 (you can use Chrome's command-line parameter -- proxy server=127.0.0.1:5000), and then accessing all other websites through it. When the browser accesses the HTTP protocol, forwarding is not a problem, but when accessing the HTTPS protocol, the target address cannot be accessed. An error message appears:

Yarp ReverseProxy Forwarder HttpForwarder [48]

RequestCreation: An error was accounted for while creating the request message

System NotSupportedException: Unsupported request method 'CONNECT'

Can you share the final solution if you have already solved it? Thank you very much!
